Fix auth URL

myw opened the pull request:
The way that the auth url was being used by AuthorizationService was too restrictive
and made assumptions about the format of the URL that are not part of the standard and are violated by some existing authorization schemes: GitLab's authorization, for example, uses a different URL format (no `v2/token` path) and does not name the `service` GET parameter sent to the authorization URL with the same format as DockerHub does (it uses`container_registry`, instead of the registry host name).

The changes in this PR do two main things:
1. It adds a parameter to the client constructor, `auth_service_name`, which allows client consumer code to provide custom service names; if not provided, the code defaults to the previous behavior (using the registry hostname).
2. It adds a new auth service URL parameter `auth_service_url_full`. In this parameter, the `/v2/token` path is no longer automatically assumed on all URLs and must be provided explicitly if needed. **This new parameter deprecates the older `auth_service_url` parameter, which will now raise a `DeprecationWarning` if provided**.

As a minor additional improvement, the PR also makes the `docker-registry-show.py` script releasted to users' PATHs upon package installation (via the `scripts` argument in `setup.py`).

myw committed:
Fix auth URL

The way that the auth url was being used by AuthorizationService was too restrictive
and made too many assumptions about the format of the URL; GitLab, for example, uses
a different URL format and does not name the service the same as DockerHub does.

These changes add a parameter, 'auth_service_name', which allows custom service names
that are not simply the same as the registry's host name (e.g. GitlLab uses
'container_registry'), and also changes the URL so that the 'v2/token' location is
not necessarily assumed.

*This is necessarily a breaking change for existing consumers and will require a majorversion bump*.

myw commented:
hello @graingert — would you be able to review and merge in this PR if it's acceptable?

myw committed:
Set a null scope on check_status

graingert commented on docker-registry-show.py:
```diff
@@ -39,8 +39,14 @@ def __init__(self):
         self.parser.add_argument('--username', metavar='USERNAME')
         self.parser.add_argument('--password', metavar='PASSWORD')
 
-        self.parser.add_argument('--authorization-service', metavar='AUTH_SERVICE', type=str,
```
let's keep backwards compatibility by adding a deprecated `--authorization-service` that automatically fills in the authorization_service_url and authorization_service_name

myw replied:
sure, that's a good idea. will do

graingert commented:
can you add a test case for this change?

myw committed:
Remove username/pw access from registry requests

This prevents the access token from being sent by the requests library

myw committed:
Revert incorrect URL update in mock_registry

myw committed:
Clean up _BaseClient module initialization

myw committed:
Maintain backward compatability and add warnings

Rather than forcing command-line and API consumers to change their code, we keep the old arguments
and behavior in, but add a warning, allowing the new behavior to be triggered with new arguments

myw committed:
Add docker-registry-show.py to scripts

This puts it in the user's PATH upon package installation

myw committed:
Clean up lint errors in conftest.py

myw committed:
Fix unit test failures from kwargs.pop bugs

myw committed:
Clean up integration test

Put errors checking before status checking

myw committed:
Add unit tests and set the warning type

myw commented:
@graingert thanks for your prompt and helpful review!  In my next round of changes, I added backwards-compatible interfaces, both at the API and command-line level, with deprecation warnings; I also added unit tests to check the functionality, and did some other minor cleanup. I'm updating the PR description to reflect that it does a deprecation, rather than a backwards-incompatible change.

Currently, the base client integration test fails for me on `put_manifest` in both python 2.7 and 3.6: the registry responds with HTTP 500 for the PUT request. However, this is also happening for me in `master` as well, so I do not think this PR introduces the failure.

Please let me know if there are any further changes you'd like me to make for this PR—I'd be happy to discuss.

On a related note, I will likely add the capability for the BaseClient to make [cross-repository blob mounts](https://docs.docker.com/registry/spec/api/#cross-repository-blob-mount) in an upcoming PR. It should be straightforward enough, and enables efficiently "moving" Docker images on a registry from one repo to another without downloading, retagging, and uploading the entire image.

armstrongli commented:
@graingert  any update on this request?

graingert commented:
I don't have permissions to merge PRs

djmattyg007 commented:
Hello. Since this package appears to be abandoned, I forked it and made several improvements. This includes dropping support for v1 registries, and overhauling how the auth service works.

If you're still feeling up to it, I'd be happy to review an updated PR against my fork:

https://github.com/djmattyg007/dreg-client

myw commented:
@djmattyg007 sure;seems easy enough; i’ll take a look when I get back to my laptop; if you haven’t heard from me or seen a PR by 9/13, feel free to ping me again.

djmattyg007 commented:
@myw ping :slightly_smiling_face:

djmattyg007 commented:
Hi @myw just checking in to ask if you'd had a chance to look at dreg-client yet
